### DISCUSSION TITLE
Best strategy for backfilling Prometheus data from backup and live data

### DISCUSSION THREAD

**Discussion by ginkel** [grafana/mimir#6977] [Help and support] [answered]
Hi there,

I set up a Mimir cluster and started backfilling data from Prometheus backups after pre-processing them with the Thanos CLI. This worked for all backed up data, which I imported in chronological order. Leaves the last month as a remainder, which I did not reconstruct from backups, but would import from the current Prometheus TSDB state.

I have not yet activated remote write for the tenant in question.

Is it ok to activate remote-write first and then fill the ~ 1 month gap using backfill or will Mimir complain that time series are not in chronological order? I'd like to avoid a gap between the part reconstructed from TSDB files and the "live" part built by remote writes...

Thanks,
Thilo

**Comment by dimitarvdimitrov** [comment, marked as answer]
>Is it ok to activate remote-write first and then fill the ~ 1 month gap using backfill or will Mimir complain that time series are not in chronological order? 

This is perfectly ok. Since the data is being backfilled directly in the object store ordering doesn't matter.

> I'd like to avoid a gap between the part reconstructed from TSDB files and the "live" part built by remote writes...

The gap will be there until you manage to backfill. After that backfilled blocks will be discovered and there won't be a gap. It might take up to 1h to discover the blocks and be able to query them after you've backfilled them.; it's usually less than 30m. Mind that you might have to flush the query-frontend results cache after the blocks are discovered. If you're using memcached, the easiest way to do that is to restart the processes.

**Reply by ginkel** [reply]
Excellent, thanks for the speedy reply! The time gap until the cache expires won't be an issue.

Will it cause problems if there is a (small) overlap between remote writes and the TSDB import?

**Reply by dimitarvdimitrov** [reply]
> Will it cause problems if there is a (small) overlap between remote writes and the TSDB import?

Data will still be returned without errors. If timestamps and samples are exactly the same (same prometheis that created the blocks are remote-wrote), then there won't be any issues.

If samples for the exact timestamp are different, then one will be picked "at random" for each query.

If timestamps misalign it's possible to get ghost counter resets for this time period. Counter resets end up causing large spikes in queries that use `increase` or `rate` for example (see https://github.com/grafana/mimir/issues/5198 for examples)
